tools/jdbc/src/main/java/org/duckdb/DuckDBArrowResult.java

```diff
@@ -0,0 +1,91 @@
+package org.duckdb;
+
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.Iterator;
+
+public class DuckDBArrowResult implements Iterable<Object> {
+
+	private Class<?> vector_root_class;
+	private Class<?> schema_class;
+	private Class<?> dictionary_class;
+
+	private Class<?> root_allocator_class;
+	private Class<?> buffer_allocator_class;
+
+	private Class<?> c_schema_class;
+	private Class<?> c_array_class;
+	private Class<?> c_data_class;
+	private Class<?> c_dictionary_class;
+
+	private Object allocator;
+
+	private DuckDBResultSet result_set;
+	private Object schema;
+	private Object vector_schema_root;
+
+	public DuckDBArrowResult(ResultSet result_set_p) throws Exception {
+		if (!(result_set_p instanceof DuckDBResultSet)) {
+			throw new SQLException("This only works with a DuckDB ResultSet");
+		}
+		result_set = (DuckDBResultSet) result_set_p;
+
+		vector_root_class = Class.forName("org.apache.arrow.vector.VectorSchemaRoot");
+		schema_class = Class.forName("org.apache.arrow.vector.types.pojo.Schema");
+		dictionary_class = Class.forName("org.apache.arrow.vector.dictionary.DictionaryProvider");
+
+		root_allocator_class = Class.forName("org.apache.arrow.memory.RootAllocator");
+		buffer_allocator_class = Class.forName("org.apache.arrow.memory.BufferAllocator");
+
+		c_schema_class = Class.forName("org.apache.arrow.c.ArrowSchema");
+		c_array_class = Class.forName("org.apache.arrow.c.ArrowArray");
+		c_data_class = Class.forName("org.apache.arrow.c.Data");
+		c_dictionary_class = Class.forName("org.apache.arrow.c.CDataDictionaryProvider");
+
+		allocator = root_allocator_class.getConstructor().newInstance();
+
+		Object c_schema = c_schema_class.getMethod("allocateNew", buffer_allocator_class).invoke(null, allocator);
+		result_set.arrowExportSchema((Long) c_schema_class.getMethod("memoryAddress").invoke(c_schema));
+		schema = c_data_class.getMethod("importSchema", buffer_allocator_class, c_schema_class, c_dictionary_class)
```

Comment:
There might be an easier / more directly supported approach than this in `org.apache.arrow.c.Data`

https://github.com/apache/arrow/blob/master/java/c/src/main/java/org/apache/arrow/c/Data.java#L392-L394

Perhaps the strategy here is more compatible with the way that you can pull results from duckdb, but using an ArrayStreamReader (and perhaps returning it as an object to not leak the types) might prove more idiomatic from the arrow side. 

However returning an iterator of VectorSchemaRoots is pretty good too (reminiscent of the JDBC -> Arrow adapter's ArrowVectorIterator).

Reply:
Let me try the stream approach

Reply:
Indeed returning an `ArrowReader` is much nicer. Only drawback is that this requires Arrow dev for now.

Reply:
@hannes I was able to get this running on the "released" 9.0.0, but I had to build arrow-c-data from source myself. 

I believe that I was working on getting support for arrow working in calcite before an ran into something similar with gandiva that also effects arrow-c-data. See error report here: https://issues.apache.org/jira/browse/ARROW-16608.